### DISCUSSION TITLE
Can JJ.exe be used as a portable application?

### DISCUSSION THREAD

**Discussion by Gavin-Holt** [jj-vcs/jj#204] [Q&A] [answered]
Hi,

My life is on a USB stick and I don't have admin rights at work (UK Hospital - Windows 10).

As a hobby programmer I do not want to install git and fight with paths/drive letters or give it 47MB.

JJ as a single file application is very attractive. Many thanks for supplying binaries and please keep the native backend.

I worked through the Tutorial:

1.  Without any installation I have used ```jj git clone``` to create a local copy of my work stored in GitHub.

2.  After editing a few files I can run ```jj st``` which reports the activity and ```jj diff``` to see the alterations.

3.  Closing the commit with ```jj close``` reports a new working copy in green - not blue.

4.  Then ```jj st``` closes the parent (in blue) and reports a new clean working copy (in green).

**How do I push my changes back to GitHub?**

*   I must admit I have not invested in learning git so I may be missing a step.

*   I have tried to run ```jj git push``` but this reports ```Nothing changed```.

*   Making a ~\AppData\Roaming\jj\config.toml file with user name and email did not help.

I am hoping for a really simple clone/pull/push workflow to get me started with GitHub.

Sorry if this is a trivial noob question.

Kind Regards Gavin Holt

**Comment by martinvonz** [comment, marked as answer]
> please keep the native backend.

Yes, that's not going away. I really don't recommend it for anything other than experiments at this point, though. It should work fine for small projects, but if you want a format that's going to be unchanged next year, you should use the git backend.

> How do I push my changes back to GitHub?

Git requires a branch, and the `jj git push` command doesn't currently have a way of creating that branch on the fly. Instead, you'll have to do `jj branch <branch name> -r <commit where it should go, typically "@-" for the parent commit>`. It's possible that `<branch name>` is `main` or `master` in your case, if this is your own repo with a single branch

> I must admit I have not invested in learning git so I may be missing a step.

Interesting. What source control system have you used so far? I thought everyone used Git these days.

**Reply by martinvonz** [reply]
Oh, and to answer the question in the title: Yes, `jj.exe` can be used as a portable application. I don't have access to a Windows machine myself, but it *should* work (i.e. feel free to file bug reports if it doesn't)

**Reply by Gavin-Holt** [reply]
Hi

Many thanks for your help. 

I am not a software professional but know enough to feel guilty not using some form of VCS for my projects.

My repo is only a main branch - made using drag and drop! https://github.com/Gavin-Holt/SciTEwithStrips

I used the ```jj branch main``` command and then ```jj git push``` on a closed commit which returned:

**Error: Unexpected git error when pushing: request failed with status code: 401;** 

This may be due to restrictions at work - I will try once I get home tonight.

Just thinking through logically - how does JJ transmit my credentials to GitHub to allow commits?

Kind Regards Gavin Holt

**Comment by martinvonz** [comment]
> I am not a software professional

Ah, now it makes sense that you have not used Git.

> Error: Unexpected git error when pushing: request failed with status code: 401;

That indicates a problem with authentication, as you seem to have guessed. jj currently only supports password-less SSH keys and `ssh-agent` (specifically in `~/.ssh/id_rsa`).

**Reply by Gavin-Holt** [reply]
Hi

Thanks for the clarification. 

I have spent some time learning about OpenSSH:

*	I have a key pair in %USERPROFILE%\.ssh\id_rsa and %USERPROFILE%\.ssh\id_rsa.pub. 
*	I have uploaded the public key to my GitHub account.

Am I right that the ssh-agent service must be running for jj to access these files? 

*	Unfortunately this will stop "portable" use for non-admin Windows users 
*	You need admin rights to launch ssh-agent.exe 
*	Every updates reset the service to ```disabled``` - don't you just love Windows!!
*	[Stackoverflow advice](https://stackoverflow.com/questions/52113738/starting-ssh-agent-on-windows-10-fails-unable-to-start-ssh-agent-service-erro)

At home with ssh-agent running the ```jj.exe git push``` still fails with:

```
Working copy now at: 479bea7726f4 (in green)                                           
Press any key to continue . . .
Error: Unexpected git error when pushing: request failed with status code: 401; class=Http (34)
```

Not quite sure what I am doing wrong. 

I am very grateful for your help and most of the above is information for others who may be hoping for a "portable" git client. 

I don't expect any changes your secure workflow to address these problems. 

Kind Regards Gavin Holt

**Reply by martinvonz** [reply]
> Am I right that the ssh-agent service must be running for jj to access these files?

Yes, if the key is password-protected, you need to run ssh-agent. Password-less key files don't need ssh-agent.

> Error: Unexpected git error when pushing: request failed with status code: 401; class=Http (34)

Just a guess, but maybe that's https://github.com/libgit2/libgit2/issues/4958. Sorry :(

> I don't expect any changes your secure workflow to address these problems.

We need to improve the authentication in general (there's #63 for one aspect of it).
